### PR TITLE
feat(view): adds lock functionality to `PreviewPanel`

### PR DESCRIPTION
**Pull request by namjul**
# Pull Request Checklist

If you are a community contributor, copy and paste the PR template from [Dendron Community PR Checklist](https://gist.github.com/kevinslin/5d1a6663638259e7dbd33b01975cc00f) and add it to the body of the pull request. 

If you are a team member, copy and paste the template from [Dendron Extended PR Checklist](https://gist.github.com/kevinslin/dfc7a101f21c58478216aa0d70256bc2).

To copy the template, click on the `Raw` button on the gist to copy the plaintext version of the template to this PR.

### CODE REVIEW

File: packages/dendron-plugin-views/src/components/DendronApp.tsx [hunk 1]
```diff
@@ -178,7 +179,29 @@ function DendronVSCodeApp({ Component }: { Component: DendronComponent }) {
     }
   });
 
-  return <Component {...props} />;
+  const isLocked = props.ide.isLocked;
+
+  const handleLock = React.useCallback(() => {
```
Comment (hkievet) [inline packages/dendron-plugin-views/src/components/DendronApp.tsx:184]: Why are you using `useCallback` here?

Comment (hkievet) [reply packages/dendron-plugin-views/src/components/DendronApp.tsx:184]: I that that you just copied this from the Preview component, but I think just defining a simple non wrapped function (yes, that gets redefined every render) might be more apt.

https://kentcdodds.com/blog/usememo-and-usecallback

TLDR, this function call is neither expensive nor makes an expensive calculation.

Comment (namjul) [reply packages/dendron-plugin-views/src/components/DendronApp.tsx]: Thank you for that pointer. I had it previously inside `DendronNotePreview` thinking about "Referential equality" but also there I oversaw its uselessness. Makes me reflect on unconsicous assumptions or automatic responses I find myself inhabiting. [`1834f16` (#3293)](https://github.com/dendronhq/dendron/pull/3293/commits/1834f167e703eb4b02d1da67432a5c5f3f8a8172)

Comment (hkievet) [reply packages/dendron-plugin-views/src/components/DendronApp.tsx]: I love that you linked the commit you made the change on.  Wow!  I've never seen that before and it's so smart.

File: packages/common-frontend/src/features/ide/slice.ts [hunk 2]
```diff
@@ -31,6 +31,7 @@ type InitialState = {
   showBacklinks?: boolean;
   showOutwardLinks?: boolean;
   showHierarchy?: boolean;
+  isLocked?: boolean;
```
Comment (namjul) [inline packages/common-frontend/src/features/ide/slice.ts:34]: We could also convert the state into local-state inside `DendronApp` as long we only need it there. Maybe premature to put it already inside the global state.

File: packages/plugin-core/src/constants.ts [hunk 3]
```diff
@@ -902,6 +902,12 @@ export const DENDRON_COMMANDS: { [key: string]: CommandEntry } = {
     },
     when: "dendron:pluginActive",
   },
+  LOCK_PREVIEW: {
+    key: "dendron.lockPreview",
+    title: `${CMD_PREFIX} Lock Preview`,
+    icon: `$(open-preview)`,
+    when: "dendron:pluginActive",
+  },
```
Comment (namjul) [inline packages/plugin-core/src/constants.ts:910]: FYI: I did not add shortkeys.

**Comment by kevinslin** [comment]
looks like some of the integration tests are breaking because our mock doesn't have the new methods. could you take a look?

File: packages/common-frontend/src/features/ide/slice.ts (hunk 2 above)
Comment (kevinslin) [reply packages/common-frontend/src/features/ide/slice.ts:34]: fine for now. we can simplify later if needed

File: packages/plugin-core/src/constants.ts (hunk 3 above)
Comment (kevinslin) [reply packages/plugin-core/src/constants.ts:910]: thats fine - we've been avoiding them for new commands

File: packages/plugin-core/src/constants.ts [hunk 4]
```diff
@@ -902,6 +902,12 @@ export const DENDRON_COMMANDS: { [key: string]: CommandEntry } = {
     },
     when: "dendron:pluginActive",
   },
+  LOCK_PREVIEW: {
+    key: "dendron.lockPreview",
+    title: `${CMD_PREFIX} Lock Preview`,
+    icon: `$(open-preview)`,
```
Comment (kevinslin) [inline packages/plugin-core/src/constants.ts:908]: can we get a different icon? see https://code.visualstudio.com/api/references/icons-in-labels

File: packages/plugin-core/src/commands/LockPreview.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  constructor(previewPanel: PreviewProxy | undefined) {
+    super();
+    this._panel = previewPanel;
+  }
+
+  async sanityCheck() {
+    if (!this._panel || !this._panel.isVisible()) {
+      return "No preview currently open";
+    }
+    if (this._panel && this._panel.isLocked()) {
```
Comment (kevinslin) [inline packages/plugin-core/src/commands/LockPreview.ts:27]: can we implement this as toggling behavior? rename to toggle preview lock

File: packages/dendron-plugin-views/src/components/DendronApp.tsx [hunk 6]
```diff
@@ -161,14 +166,45 @@ function DendronVSCodeApp({ Component }: { Component: DendronComponent }) {
...
+      type: NoteViewMessageEnum.onUnlock,
+      data: {},
+      source: DMessageSource.webClient,
+    });
+  };
+
+  return (
+    <>
+      <Component {...props} />
+      {isLocked && (
```
Comment (kevinslin) [inline packages/dendron-plugin-views/src/components/DendronApp.tsx:198]: to make this discoverable, can we have the lock be transparent if not locked so that users can toggle by clicking on it? its okay to split this out as a separate pr

**Review by kevinslin — CHANGES_REQUESTED** [review]
nice job! left some ux comments

File: packages/plugin-core/src/constants.ts (hunk 4 above)
Comment (namjul) [reply packages/plugin-core/src/constants.ts]: [`75aa659` (#3293)](https://github.com/dendronhq/dendron/pull/3293/commits/75aa6598d30ca0be5dc4efdb3ebef196d764727a)

File: packages/plugin-core/src/commands/LockPreview.ts (hunk 5 above)
Comment (namjul) [reply packages/plugin-core/src/commands/LockPreview.ts]: [`2ad67d2` (#3293)](https://github.com/dendronhq/dendron/pull/3293/commits/2ad67d2337df532b7348c34e12eeab5344c91c4f)
[`e827c0f` (#3293)](https://github.com/dendronhq/dendron/pull/3293/commits/e827c0f0d458712e63645f43a3bbbd17b7cf7fa9)

File: packages/plugin-core/src/components/views/PreviewPanel.ts [hunk 7]
```diff
@@ -137,7 +138,7 @@ export class PreviewPanel implements PreviewProxy, vscode.Disposable {
           this._onTextChanged = undefined;
         }
 
-        this._panel = undefined;
+        this.dispose();
```
Comment (namjul) [inline packages/plugin-core/src/components/views/PreviewPanel.ts:141]: @kevinslin not super confident about this change. Maybe you can check if this does not break existing behavior.

File: packages/dendron-plugin-views/src/components/DendronApp.tsx (hunk 6 above)
Comment (namjul) [reply packages/dendron-plugin-views/src/components/DendronApp.tsx:198]: will catch up on that here https://github.com/dendronhq/dendron/tree/enhance/toggle-preview-lock-from-inside-preview

File: packages/plugin-core/src/components/views/PreviewPanel.ts (hunk 7 above)
Comment (kevinslin) [reply packages/plugin-core/src/components/views/PreviewPanel.ts:141]: this shouldn't be necessary since this is invoked after `panel.dispose` is called

Comment (namjul) [reply packages/plugin-core/src/components/views/PreviewPanel.ts:141]: ok so you are pointing to `this._panel = undefined;` being called in  `dispose()` meaning we are confident about that change and can leave it like it is.

Comment (kevinslin) [reply packages/plugin-core/src/components/views/PreviewPanel.ts:141]: if by leaving as is you mean reverting this change (not adding `this.dispose`), than yes. `this.dispose` invokes `this._panel.onDidDispose` 
adding @jonathanyeung in case he has more to add here

File: packages/plugin-core/src/components/views/PreviewViewFactory.ts [hunk 8]
```diff
@@ -30,4 +30,8 @@ export class PreviewPanelFactory {
 
     return PreviewPanelFactory._preview;
   }
+
+  static get preview() {
+    return PreviewPanelFactory._preview
```
Comment (jonathanyeung) [inline packages/plugin-core/src/components/views/PreviewViewFactory.ts:35]: Let's not add a static getter for `PreviewProxy`, otherwise it'll be hard to manage coupling around the preview panel later (we had this problem in the past).  Instead, in `_extension.ts` you can create a single object and share it between the implementations of the TOGGLE_PREVIEW_LOCK and TOGGLE_PREVIEW commands

**Review by jonathanyeung — CHANGES_REQUESTED** [review]
Looks cool!

File: packages/plugin-core/src/components/views/PreviewPanel.ts (hunk 7 above)
Comment (namjul) [reply packages/plugin-core/src/components/views/PreviewPanel.ts]: @jonathanyeung the intention with this commit was to go after behavoir 1.

[`d13cef8` (#3293)](https://github.com/dendronhq/dendron/pull/3293/commits/d13cef8231ee2c9e8bc39ce2a597a31ed38b54dc)

@kevinslin after some [delayering/untangling](https://twitter.com/Malcolm_Ocean/status/1363299005443108864) my confusion was about not seing `dispose` and `onDidDispose` being two different entry points. obvious now :)

File: packages/plugin-core/src/components/views/PreviewPanel.ts [hunk 9]
```diff
@@ -150,13 +151,42 @@ export class PreviewPanel implements PreviewProxy, vscode.Disposable {
   hide(): void {
     this.dispose();
   }
+  lock() {
+    const activeTextEditor = VSCodeUtils.getActiveTextEditor();
+    if (activeTextEditor) {
+      this._lockedEditorFileName = activeTextEditor?.document.fileName;
+      this.sendLockMessage(this._panel!, this.isLocked());
+    } else {
+      vscode.window.showInformationMessage("No active texteditor found");
```
[the remark this answers is not in the document]
Comment (namjul) [reply packages/plugin-core/src/components/views/PreviewPanel.ts]: Good point, since we do the sanittyCheck in `packages/plugin-core/src/commands/TogglePreviewLock.ts 24` we should not enter this path.

[`7485b1a` (#3293)](https://github.com/dendronhq/dendron/pull/3293/commits/7485b1a7f3e8176005b058039d8fc5d60da73866)

File: packages/plugin-core/src/components/views/PreviewViewFactory.ts (hunk 8 above)
Comment (namjul) [reply packages/plugin-core/src/components/views/PreviewViewFactory.ts]: I was uncomfortable with this change, but did not take time to think about it more. Thank you for pointing this out, thats a good way to solve this.

[`b147ea1` (#3293)](https://github.com/dendronhq/dendron/pull/3293/commits/b147ea11f4742cd0a18334c7bc32913092d405b1)

File: packages/plugin-core/src/components/views/PreviewPanel.ts [hunk 10]
```diff
@@ -346,6 +387,7 @@ export class PreviewPanel implements PreviewProxy, vscode.Disposable {
           data: {
             note,
             syncChangedNote,
+            isLocked: this.isLocked(),
```
[the remark this answers is not in the document]
Comment (namjul) [reply packages/plugin-core/src/components/views/PreviewPanel.ts]: [`64ae6c4` (#3293)](https://github.com/dendronhq/dendron/pull/3293/commits/64ae6c48bd50fbe461b76a7e42a534f65c93f5cf)

agreed. this is not necessary now since we go for behavoir 1 (reseting lock state after dispose). If we conclude behavior 2 at some point we will need to sync this up with the webview panel but then could just call `sendLockMessage` in `packages/plugin-core/src/components/views/PreviewPanel.ts 149`

**Review by kevinslin — APPROVED** [review]

**Review by jonathanyeung — APPROVED** [review]
